homeassistant/components/template/fan.py

```diff
@@ -280,17 +357,53 @@ async def async_turn_off(self) -> None:
 
     async def async_set_speed(self, speed: str) -> None:
         """Set the speed of the fan."""
-        if self._set_speed_script is None:
+        if speed not in self.speed_list:
+            _LOGGER.error(
+                "Received invalid speed: %s. Expected: %s", speed, self.speed_list
+            )
             return
 
-        if speed in self._speed_list:
-            self._speed = speed
+        self._state = STATE_OFF if speed == SPEED_OFF else STATE_ON
+        self._speed = speed
+        self._preset_mode = None
+        self._percentage = self.speed_to_percentage(speed)
+
+        if self._set_speed_script:
             await self._set_speed_script.async_run(
-                {ATTR_SPEED: speed}, context=self._context
+                {ATTR_SPEED: self._speed}, context=self._context
             )
-        else:
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the percentage speed of the fan."""
+        speed_list = self.speed_list
+        self._state = STATE_OFF if percentage == 0 else STATE_ON
+        self._speed = self.percentage_to_speed(percentage) if speed_list else None
+        self._percentage = percentage
+        self._preset_mode = None
+
+        if self._set_percentage_script:
+            await self._set_percentage_script.async_run(
+                {ATTR_PERCENTAGE: self._percentage}, context=self._context
+            )
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset_mode of the fan."""
+        if preset_mode not in self.preset_modes:
             _LOGGER.error(
-                "Received invalid speed: %s. Expected: %s", speed, self._speed_list
+                "Received invalid preset_mode: %s. Expected: %s",
+                preset_mode,
+                self.preset_modes,
+            )
+            return
+
+        self._state = STATE_ON
```

Comment:
Out of scope for this PR, but we can make `self._state` a boolean. It is compared to `STATE_ON` inside `is_on` property.